### ISSUE TITLE
Anomaly "grounding a non evar-free term"

### ISSUE THREAD

**Issue by tchajed** [coq/coq#8224]
#### Version

Coq 8.9+alpha, commit 18b662aa30

#### Operating system

macOS 10.13.6

#### Description of the problem

The following code raises an anomaly:

```
Record t A :=
    { X : option A; }.

Check {| X := fun _ => _ |}.
```

**Comment by gares** [comment]
This is a meta-issue to be discussed with @ejgallego .

It comes from a patch of his that strengthens an invariant. The patch was tested on code that works.
You example should fail, and I guess is the code generating the error message that breaks the invariant.

I think we may keep the invariant active in the development phase, but don't release Coq with the invariant check.

**Comment by tchajed** [comment]
Ok, that explains why I've started seeing this error message (especially with the guardedness checker). I'd still like to have this broken code in the test suite. I pretty much exclusively use Coq master and when I ran into this it was annoying to not get a proper type mismatch error.

**Comment by ejgallego** [comment]
I didn't comment because the fix has been ready for a long time, it has been not just merged, see the relevant issues.

**Comment by herbelin** [comment]
Another instance (which is not fixed by #8427):
```coq
Fixpoint restrict f n :=
  match n with
  | O => nil
  | S n => cons (f n) (restrict f n)
  end.
```
It might indeed be wise to be more tolerant on this invariant check in the final 8.9.
